beacon_node/lighthouse_network/src/peer_manager/mod.rs

```diff
@@ -819,26 +833,290 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
+                                        continue;
+                                    }
+                                }
+
+                                // Check the sync committee
+                                if let Some(subnets) = peer_to_sync_committee.get(candidate_peer) {
+                                    // The peer is subscribed to some long-lived sync-committees
+                                    // Of all the subnets this peer is subscribed too, the minimum
+                                    // peer count of all of them is min_subnet_count
+                                    if let Some(min_subnet_count) = subnets
+                                        .iter()
+                                        .map(|v| {
+                                            *sync_committee_peer_count.get(v).unwrap_or(&100u64)
+                                        })
+                                        .min()
+                                    {
+                                        // If the minimum count is our target or lower, we
+                                        // shouldn't remove this peer, because it drops us lower
+                                        // than our target
+                                        if min_subnet_count <= MIN_SYNC_COMMITTEE_PEERS {
+                                            // Do not drop this peer in this pruning interval
+                                            continue;
+                                        }
+                                    }
+                                }
+
+                                // This peer is suitable to be pruned
+                                removed_peer_index = Some(index);
+                                break;
+                            }
+
+                            // If we have successfully found a candidate peer to prune, prune it,
+                            // otherwise all peers on this subnet should not be removed due to our
+                            // outbound limit or min_subnet_count. In this case, we remove all
+                            // peers from the pruning logic and try another subnet.
+                            if let Some(index) = removed_peer_index {
+                                let (candidate_peer, _) = peers_on_subnet.remove(index);
+                                // Remove pruned peers from other subnet counts
+                                for subnet_peers in subnet_to_peer.values_mut() {
+                                    subnet_peers.retain(|(peer_id, _)| peer_id != &candidate_peer);
+                                }
+                                peers_to_prune.insert(candidate_peer);
+                            } else {
+                                peers_on_subnet.clear();
+                            }
+                            continue;
+                        }
+                    } else {
+                        unreachable!();
```

Comment:
I think we can avoid this `unreachable` by using `iter_mut` in the top-level condition:

```rust
if let Some((_, peers_on_subnet)) = subnet_to_peer
    .iter_mut()
    .max_by_key(|(_, peers)| peers.len())
{
```